lib/internal/modules/cjs/loader.js

```diff
@@ -777,69 +657,173 @@ Module.runMain = function() {
+
+  let paths = [path.resolve(prefixDir, 'lib', 'node')];
 
   if (homeDir) {
     paths.unshift(path.resolve(homeDir, '.node_libraries'));
     paths.unshift(path.resolve(homeDir, '.node_modules'));
   }
 
   if (nodePath) {
-    paths = nodePath.split(path.delimiter).filter(function pathsFilterCB(path) {
-      return !!path;
-    }).concat(paths);
+    paths = nodePath.split(path.delimiter)
+                    .filter((path) => !!path)
+                    .concat(paths);
   }
 
-  modulePaths = paths;
+  // Filter all paths that do not exist or that resolve to a file instead of a
+  // directory.
+  modulePaths = paths.filter((path) => stat(path) === 1);
 
-  // clone as a shallow copy, for introspection.
-  Module.globalPaths = modulePaths.slice(0);
+  // Use the original input for introspection.
+  Module.globalPaths = paths;
 };
 
 Module._preloadModules = function(requests) {
-  if (!Array.isArray(requests))
-    return;
-
   // Preloaded modules have a dummy parent module which is deemed to exist
   // in the current working directory. This seeds the search path for
   // preloaded modules.
-  var parent = new Module('internal/preload', null);
+  const parent = new Module('internal/preload');
   try {
-    parent.paths = Module._nodeModulePaths(process.cwd());
+    parent.paths = getNodeModulePaths(process.cwd());
   } catch (e) {
     if (e.code !== 'ENOENT') {
       throw e;
     }
+    parent.paths = [];
+  }
+  for (const request of requests) {
+    parent.require(request);
   }
-  for (var n = 0; n < requests.length; n++)
-    parent.require(requests[n]);
 };
 
 Module._initPaths();
 
 // Backwards compatibility
 Module.Module = Module;
+
+const extensions = Object.create(null);
+
+// Native extension for .js
+extensions['.js'] = function(module, filename) {
+  const content = fs.readFileSync(filename, 'utf8');
+  module._compile(stripBOM(content), filename);
+};
+
+// Native extension for .json
+extensions['.json'] = function(module, filename) {
+  const content = fs.readFileSync(filename, 'utf8');
+  try {
+    module.exports = JSON.parse(stripBOM(content));
+  } catch (err) {
+    err.message = `${filename}: ${err.message}`;
+    throw err;
+  }
+};
+
+// Native extension for .node
+extensions['.node'] = function(module, filename) {
+  return process.dlopen(module, path.toNamespacedPath(filename));
+};
+
+if (experimentalModules) {
+  extensions['.mjs'] = function(module, filename) {
+    throw new ERR_REQUIRE_ESM(filename);
+  };
+}
+
+// Use proxies to optimize resetting the extensions. This would otherwise be a
+// significant overhead as we have to check for all object keys each time a
+// uncached module is loaded (everything else would be a breaking change).
+function setExtensionsProxy(extensions) {
+  // Intersect manipulating the extensions.
+  extensionNames = Object.keys(extensions);
+  extensionCopy = Object.create(null);
+  for (const name of extensionNames) {
+    extensionCopy[name] = true;
+  }
```

Comment:
It seems like an awful lot of work for to avoid `Object.keys` here.
How much is the win here for just `Object.keys`? Can we leave this whole `Proxy` of the `Module` out and still get the majority of the wins?